File: m4d/ClientApp/src/models/DanceQueryItem.ts
import { jsonMember, jsonObject } from "typedjson";
import type { NamedObject } from "./DanceDatabase/NamedObject";
import { safeDanceDatabase } from "@/helpers/DanceEnvironmentManager";
import { TagQuery } from "./TagQuery";

const primaryMarker = "*";

@jsonObject
export class DanceQueryItem {
  @jsonMember(String) public id!: string;
  @jsonMember(Number) public threshold!: number;
  @jsonMember(String) public tags?: string;
  // Marks this dance as the explicit target for dance-rating sort, tempo sort, and the
  // tempo range filter when more than one dance is selected - see DanceQuery.primaryDanceId.
  @jsonMember(Boolean) public primary?: boolean;

  public static fromValue(value: string): DanceQueryItem {
    const regex = /^([a-zA-Z0-9]+)(\*?)([+-]?)(\d*)\|?(.*)?$/;
    const match = value.match(regex);
    if (!match) {
      throw new Error(`Invalid value format: ${value}`);
    }

    const dance = safeDanceDatabase().fromId(match[1] || "");
    if (!dance) {
      throw new Error(`Couldn't find dance ${match[1]}`);
    }

    const weight = match[4] ? parseInt(match[4]) : 1;
    const tags = match[5] ?? undefined;

    return new DanceQueryItem({
      id: dance.id,
      primary: match[2] === primaryMarker ? true : undefined,
      threshold: match[3] === "-" ? -weight : weight,
      tags: tags ? tags : undefined,
    });
  }

  public constructor(init?: Partial<DanceQueryItem>) {
    Object.assign(this, init);
  }

  public get dance(): NamedObject {
    return safeDanceDatabase().fromId(this.id)!;
  }

  public get tagQuery(): TagQuery | undefined {
    if (this.tags && this.tags.length > 0) {
      return new TagQuery(this.tags);
    }
    return undefined;
  }

  public get value(): string {
    return this.toString();
  }

  public toString(): string {
    const baseStr = `${this.id}${this.primary ? primaryMarker : ""}${this.threshold !== 1 ? (this.threshold < 0 ? "-" : "+") + Math.abs(this.threshold) : ""}`;
    if (this.tags && this.tags.length > 0) {
      return `${baseStr}|${this.tags}`;
    }
    return baseStr;
  }

  public get description(): string {
    const modifiers = [];
    if (this.threshold !== 1) {
      modifiers.push(
        this.threshold > 0 ? `at least ${this.threshold}` : `at most ${Math.abs(this.threshold)}`,
      );
    }
    if (this.tagQuery?.hasTags) {
      modifiers.push(this.tagQuery.description);
    }
    let desc = this.dance.name;
    if (modifiers.length > 0) {
      desc = `${desc} (${modifiers.join(", ")})`;
    }
    return desc;
  }

  public get shortDescription(): string {
    const modifiers = [];
    if (this.threshold !== 1) {
      modifiers.push(`${this.threshold > 0 ? ">=" : "<="}${Math.abs(this.threshold)}`);
    }
    if (this.tagQuery?.hasTags) {
      modifiers.push(this.tagQuery.shortDescription);
    }
    let desc = this.dance.name;
    if (modifiers.length > 0) {
      desc = `${desc} (${modifiers.join(", ")})`;
    }
    return desc;
  }
}

File: m4dModels/DanceQueryItem.cs
using DanceLibrary;

using System.Text.RegularExpressions;

namespace m4dModels;

public partial class DanceQueryItem
{
    private const char PrimaryMarker = '*';

    public string Id { get; set; }
    public int Threshold { get; set; }
    public TagQuery TagQuery { get; set; }

    // Marks this dance as the explicit target for dance-rating sort, tempo sort, and the
    // tempo range filter when more than one dance is selected - see DanceQuery.PrimaryDanceId.
    public bool IsPrimary { get; set; }

    // Only meaningful when Id refers to a DanceGroup: a group has no per-dance rating/tempo
    // fields of its own, so the marker must name which member dance to actually use. Null for
    // a marked plain dance (it's simply its own target). See DanceQuery.PrimaryDanceId.
    public string PrimaryTargetId { get; set; }

    public bool IsSimple => Threshold == 1 && !IsPrimary && TagQuery.TagList.IsEmpty;

    public static DanceQueryItem FromValue(string value)
    {
        var regex = ThresholdWithTagsRegex();
        var match = regex.Match(value);
        if (!match.Success)
        {
            throw new Exception($"Invalid value format: {value}");
        }

        var dance = Dances.Instance.DanceFromId(match.Groups[1].Value) ?? throw new Exception($"Couldn't find dance {match.Groups[1].Value}");
        var weight = match.Groups[5].Success && !string.IsNullOrEmpty(match.Groups[5].Value)
            ? int.Parse(match.Groups[5].Value) : 1;

        var tags = match.Groups[6].Success ? match.Groups[6].Value : null;

        return new DanceQueryItem
        {
            Id = dance.Id,
            IsPrimary = match.Groups[2].Success,
            PrimaryTargetId = match.Groups[3].Success && !string.IsNullOrEmpty(match.Groups[3].Value)
                ? match.Groups[3].Value
                : null,
            Threshold = match.Groups[4].Success && match.Groups[4].Value == "-" ? -weight : weight,
            TagQuery = new TagQuery(tags)
        };
    }

    public DanceObject Dance => Dances.Instance.DanceFromId(Id);

    public override string ToString()
    {
        var marker = IsPrimary ? PrimaryMarker.ToString() + (PrimaryTargetId ?? "") : "";
        var baseStr =
            $"{Id}{marker}{(Threshold != 1 ? (Threshold < 0 ? "-" : "+") + Math.Abs(Threshold) : "")}";
        if (TagQuery?.TagList?.IsEmpty != true)
        {
            return $"{baseStr}|{TagQuery.TagList}";
        }
        return baseStr;
    }

    public string Description
    {
        get
        {
            var modifiers = new System.Collections.Generic.List<string>();
            if (Threshold != 1)
            {
                modifiers.Add(Threshold > 0
                    ? $"with at least {Threshold} votes"
                    : $"with at most {Math.Abs(Threshold)} votes");
            }
            if (TagQuery != null && !TagQuery.IsEmpty)
            {
                var separator = "";
                modifiers.Add(TagQuery.Description(ref separator));
            }
            var desc = Dance.Name;
            if (modifiers.Count > 0)
            {
                desc = $"{desc} ({string.Join(", ", modifiers)})";
            }
            return desc;
        }
    }

    public string ShortDescription
    {
        get
        {
            var modifiers = new System.Collections.Generic.List<string>();
            if (Threshold != 1)
            {
                modifiers.Add(Threshold > 0
                    ? $">={Threshold}"
                    : $"<={Math.Abs(Threshold)}");
            }
            if (TagQuery != null && !TagQuery.TagList.IsEmpty)
            {
                var separator = "";
                modifiers.Add(TagQuery.ShortDescription(ref separator));
            }
            var desc = Dance.Name;
            if (modifiers.Count > 0)
            {
                desc = $"{desc} ({string.Join(", ", modifiers)})";
            }
            return desc;
        }
    }

    [GeneratedRegex(@"^([a-zA-Z0-9]+)(\*)?([a-zA-Z0-9]+)?([+-]?)(\d*)\|?(.*)?$")]
    private static partial Regex ThresholdWithTagsRegex();
}
